Switch from flaky waitForRenderComplete to a sleep(4000)

LeeDr opened the pull request:
## Summary

waitForRenderComplete is too flaky.  It's going to cause large amounts of tests to be skipped if we don't do something.  The core of the problem is when we do something in a dashboard and trying to wait just until each visualization sets its data-render-complete attribute to true.  But there are issues where the saved search data table doesn't set it.  There used to be an issue with timelion.  I'm not sure if that one is 100% fixed or still could be flaky.
Until we have a rock solid method for waiting for all visualizations to render we should just sleep a few seconds.

Failing test: UI Functional Tests.test/functional/apps/dashboard/_embeddable_rendering·js - dashboard app using current data dashboard embeddable rendering adding visualizations
Fixes: https://github.com/elastic/kibana/issues/28818

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.


~~- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
~~- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
~~- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
~~- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios~~
~~- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

~~- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
~~- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11992/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12010/)

LeeDr commented:
So it turns out there were 2 potential problems which at first glance look to be the same.  One problem is in waitForRenderComplete where some visualizations don't (or don't always) set `data-render-complete="true"`.  I replaced this check with a sleep.

The other problem is sometimes not all visualizations get added.  I added logging to the method `addEveryEmbeddableOnCurrentPage()` so that it logs the names of each visualization and I also added a 100ms sleep.  I'm not really sure that's needed.  Only many test runs will tell.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12136/)

LeeDr commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12148/)

Stacey-Gammon commented:
I am investigating a fix for this here: https://github.com/elastic/kibana/pull/30419, with no sleeps.  I'd rather leave the tests off than fix it with sleeps.

LeeDr commented:
x-pack-ciGroup7 failed with something like an out of memory error;

```
00:12:46.738  info [  kibana  ] Running optimizer
00:12:46.738    │ debg Temporarily installing node.js for linux-x64
00:12:50.238    │ info Running bin/kibana to trigger the optimizer
00:12:50.238    │ debg $ ./bin/kibana --env.name=production --logging.json=false --optimize
00:12:51.700    │ debg   log   [04:52:36.734] [info][plugins-service] Plugin initialization disabled.
00:13:10.286    │ debg   log   [04:52:55.323] [info][optimize] Optimizing and caching bundles for graph, monitoring, space_selector, login, logout, logged_out, ml, dashboardViewer, apm, maps, canvas, infra, uptime, kibana, stateSessionStorageRedirect, status_page and timelion. This may take a few minutes
00:13:10.511    │ debg   log   [04:52:55.543] [info][optimize:dynamic_dll_plugin] Started dynamic dll plugin tasks
00:15:35.209    │ debg   log   [04:55:20.246] [info][optimize:dynamic_dll_plugin] Need to compile the client vendors dll
00:15:35.214    │ debg   log   [04:55:20.257] [info][optimize:dynamic_dll_plugin] Client vendors dll compilation started
00:17:15.132    │ debg 
00:17:15.132    │ debg <--- Last few GCs --->
00:17:15.132    │ debg 
00:17:15.132    │ debg [5776:0x35fb0d0]    83508 ms: Scavenge 1313.3 (1419.6) -> 1312.5 (1420.1) MB, 2.8 / 0.0 ms  (average mu = 0.316, current mu = 0.272) allocation failure 
00:17:15.133    │ debg [5776:0x35fb0d0]    83514 ms: Scavenge 1313.4 (1420.1) -> 1312.6 (1420.6) MB, 3.1 / 0.0 ms  (average mu = 0.316, current mu = 0.272) allocation failure 
00:17:15.133    │ debg [5776:0x35fb0d0]    83519 ms: Scavenge 1313.5 (1420.6) -> 1312.7 (1421.1) MB, 2.9 / 0.0 ms  (average mu = 0.316, current mu = 0.272) allocation failure 
00:17:15.133    │ debg 
00:17:15.133    │ debg 
00:17:15.133    │ debg <--- JS stacktrace --->
00:17:15.133    │ debg 
00:17:15.133    │ debg ==== JS stack trace =========================================
00:17:15.133    │ debg 
00:17:15.133    │ debg     0: ExitFrame [pc: 0x23848ac5be1d]
00:17:15.136    │ debg Security context: 0x2ba6ec31e6e1 <JSObject>
00:17:15.136    │ debg     1: /* anonymous */ [0x330bf909751] [/var/lib/jenkins/workspace/elastic+kibana+pull-request/JOB/x-pack-ciGroup7/node/immutable/kibana/build/kibana/node_modules/terser/dist/bundle.js:~44] [pc=0x23848b1eea7d](this=0x025f7b577029 <Object map = 0x224c18562189>,e=0x209c51db00e9 <AST_SymbolRef map = 0x224c185603f9>,n=0x374f0b5026f1 <undefined>)
00:17:15.136    │ debg     2: arguments...
00:17:15.136    │ debg FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
00:17:15.136    │ debg  1: 0x8dc1c0 node::Abort() [/var/lib/jenkins/workspace/elastic+kibana+pull-request/JOB/x-pack-ciGroup7/node/immutable/kibana/build/kibana/node/bin/node]

```
retest

LeeDr commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12198/)

LeeDr commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12200/)

dmlemeshko commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12218/)

markov00 commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12226/)

Stacey-Gammon commented:
Could we please give https://github.com/elastic/kibana/pull/30419 a shot first?  My suspicion is that 1) the retry is too short for some of the things it's waiting for (e.g. waiting for a search table to complete rendering), and b) we check the panel count too quickly and fail immediately, but if we wrap it in a retry so it continues to check over and over, we'll make sure it's not that the panel just hasn't finished loading on the screen yet, but legitimately isn't there.  This is my approach with: https://github.com/elastic/kibana/pull/30419.  I'm also spitting out the specific panels that have failed to emit the render-complete=true so we will have more data to know which type of vis we are waiting on.

LeeDr commented:
@gammon I changed a couple of these tests up a bit so that instead of just checking the count of panels, it checks the titles of the panels.  I did that because I found in at least one case the dashboard didn't have all the expected panels.  Case in point, check these failed results;
https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12226/JOB=kibana-ciGroup2,node=immutable/testReport/UI%20Functional%20Tests/test_functional_apps_dashboard__embeddable_rendering%C2%B7js/dashboard_app_using_current_data_dashboard_embeddable_rendering_adding_visualizations/
and you see we didn't get one panel added for some reason;
```
[00:01:59]                 │       + expected - actual
[00:01:59]                 │ 
[00:01:59]                 │          "Rendering Test: tsvb-markdown"
[00:01:59]                 │          "Rendering Test: tsvb-topn"
[00:01:59]                 │          "Rendering Test: tsvb-metric"
[00:01:59]                 │          "Rendering Test: tsvb-ts"
[00:01:59]                 │       +  "Rendering Test: geo map"
[00:01:59]                 │          "Rendering Test: input control parent"
[00:01:59]                 │          "Rendering Test: animal sounds pie"
[00:01:59]                 │          "Rendering Test: area with not filter"
[00:01:59]                 │          "Rendering Test: scripted filter and query"
```

LeeDr commented:
retest

markov00 commented:
@LeeDr I've also see that is not always the same panel. Checking previous Jenkins build I've seen that other panel are not added like the tsvb-metric. So I think it's not a real problem from the `waitForRenderComplete` but maybe on the `addEveryVisualization` code

LeeDr commented:
@markov00 Yes, I've seen a couple of different panels missing.  I even added a 100ms sleep in `addEveryEmbeddableOnCurrentPage` and it didn't solve that problem.

Stacey-Gammon commented:
It's also possible that it takes time for the visualization to show up on the DOM, especially when clicking so many in a row.  So if you could have something like this:

```
  const panelCount = await PageObjects.dashboard.getPanelCount();
  // Fails because the panel hasn't shown up in the DOM yet.
  expect(panelCount).to.be(27);	
```

That would be fixed by something like this:

```
  const panelCount = await PageObjects.dashboard.getPanelCount();
  await sleep(4000)
  // Passes because of the extra time
  expect(panelCount).to.be(27);	
```

But instead of using the sleep, I'm continuously checking (it does this internally by wrapping in a retry):
```
  await dashboardExpect.panelCount(27);
```

so better than a sleep because it will check non stop for max retry.try time (40s now, I bumped to 2 min though), and it will fall out as soon as it's done.  That's why I prefer that method instead of the sleep.

If we think `addEveryVisualization` isn't working correctly we could also throw in something like:
```
while (more vis) {
  await addVis();
  panelCount += 1 ;
  await dashboardExpect.panelCount(panelCount);
}
```

Which would continue to avoid sleeps.

Stacey-Gammon commented on test/functional/services/dashboard/add_panel.js:
```diff
@@ -47,7 +47,7 @@ export function DashboardAddPanelProvider({ getService, getPageObjects }) {
         await embeddableRows[i].click();
         log.debug(`added ${await embeddableRows[i].getVisibleText()}`);
-        await PageObjects.common.sleep(100);
+        await PageObjects.common.sleep(1000);
```
Avoid sleeps by expecting the panel count to be i + 1.

LeeDr commented:
@Stacey-Gammon Yep, it could be that.  It could be that the visualization will appear but just hasn't yet.  I'm going to try a retry there too.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12268/)

Stacey-Gammon commented:
Another thing I think is wrong ( my PR did just fail with not the right panels being added)

```

    async filterEmbeddableNames(name) {
      // The search input field may be disabled while the table is loading so wait for it
      await this.waitForEuiTableLoading();
      await testSubjects.setValue('savedObjectFinderSearchInput', name);
      await PageObjects.header.waitUntilLoadingHasFinished();
    }
```

I think we need a second `      await this.waitForEuiTableLoading();` after the search. There is no global loading header so that last one shouldn't do anything, we need the ui table loading thing to stop.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12303/)

LeeDr commented:
This was fixed by a different PR
